### ISSUE TITLE
Create Test Environment

### ISSUE THREAD

**Issue by hkatzdev** [hackforla/tables#43]
### Overview
We need to make a proper test environment to ensure that no changes break the production table.

### Action Items
- [x] Document all the current tables' schemas.
- [x] Create a new table solely for testing.
- [X] Implement schema into table.
- [ ] Add scripts / bots to run in the testing table.
- - [ ] Look into Library
- [x] Add forms to allow test data.
- [x] Populate the table with test data.
- - [x] Add test drives
- - [x] Add test calendar
- - [x] Add test github
- [ ] remove the dependency on #12 

### Resources/Instructions
https://tables.area120.google.com/u/0/workspace/bHKoRlcCZyubeRsnie2OgH/table/aUgD85hZoYo2epgVPd4O1c
https://docs.google.com/spreadsheets/d/1kHsAi6jPpXaMpdr7YCBlJiePrzoNh6LSgDqvwO2_kec/edit
https://tables.area120.google.com/u/0/workspace/8JsRrAd7fEgd5jspaL4_qD/table/9gZ9xePJnfs2H65LCypkLZ
https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect
- email for staging tables-staging@hackforla.org

**Comment by hkatzdev** [comment]
An approximent schema is available at https://docs.google.com/spreadsheets/d/1kHsAi6jPpXaMpdr7YCBlJiePrzoNh6LSgDqvwO2_kec/edit. It doesn't have all the information necessary / some details are missing, but it's enough to implement the test table.

**Comment by hkatzdev** [comment]
Test Table is implemented at https://tables.area120.google.com/u/0/workspace/8JsRrAd7fEgd5jspaL4_qD/table/9gZ9xePJnfs2H65LCypkLZ.

**Comment by hkatzdev** [comment]
TODO: Figure out Feature label

**Comment by hkatzdev** [comment]
Without some sort of automation, implementing the exact form from production and keeping it updated seems unfeasible. Instead, the text is ignored, with only the form fields being accurate - that shouldn't change much, and if it does should be tested in the test table first. May explore automation but out of scope for entry test environment - perhaps an enhancement later?

**Comment by hkatzdev** [comment]
pnpm doesn't work, npm works
using clasp to push
GitHub secret for oauth file
cat to temp dir
if secret change libsodium GitHub API update
every 30 days run a commit and key update
core pack to ensure same npm version - no real need to dockerize
can't use workspace federation - no service account allowed
gcp out of scope
staging and main branch - tables1 and tables separate account?
give dev tables1 code.
GitHub action runs on PR to main from staging and PRs to staging for tables1.
Only staging can PR to main.
tables_env - make issue!

**Comment by ExperimentsInHonesty** [comment]
1. Progress: "What is the current status of your project? What have you completed and what is left to do?"
2. Blockers: 
    - hard coded elements
    - needs a public url and security for that.
4. Availability: "How much time will you have this week to work on this issue?"
5. ETA: "When do you expect this issue to be completed?"
6. Pictures or links* (if necessary): "Add any pictures or links that will help illustrate what you are working on."

* remember to add links to the top of the issue if they are going to be needed again.

**Comment by hkatzdev** [comment]
Zoom blocker is finished. Zoom has been implemented in a way that forgoes the need for a public url with server to server oauth.

No blockers remain, should be done by next week. Basic logic is done, refactoring should be complete but needs additional testing to verify edge cases. Test Table needs to be updated to support Zoom IDs but that is trivial. Faked environment variables are done, clasp is confirmed to work correctly.

Main things I am working on is the github action script. Current solution is to use clasp but its ugly and presenting minor complications due to it not being intended for library use. Honestly may be easier and safer from a stability standpoint to use apis directly - checking those. However, unless the APIs are very simple I will probably implement clasp for now and switch to the APIs at a later time.

Clasp allows for prettify which is wonderful. The current code style isn't consistent (partly my own fault, partly due to multiple devs previously). This may seem minor but it can lead to some confusion. Additionally, the original script had several dead functions that added unneeded complexity. I'm going to recommend we use semi-standard, though tbh I don't really care what style we use - if there is a standard style for HackForLA that I am unaware of we can use that instead.

As mentioned earlier, deployments are not necessary at this time. They may need to be implemented when we add Github OAuth, but nothing I am doing should _prevent_ that.

The main non github action thing left is automated copying of the test table if possible. This is trivial if possible. If it is not possible, then the test table will need to be able to reset itself, which is easy to do but a bit tedious to code. I don't anticipate any issues with this.

**Comment by hkatzdev** [comment]
I will do a larger writeup in the future, but a TL'DR is that we are switching to Appscript to Cloudflare Workers

I have some questions before I proceed.
1. @wesley115 do you have a Windows computer? If you do can you try to install https://github.com/Schniz/fnm? I recommend using scoop if you don't have administrator permissions, winget if you do and you don't want to install scoop, and WingetGui if you aren't used to installing packages with the command line. If it takes more than 10 minutes to figure out how to then please stop and mention how far you got and the steps you took here - I'm trying to determine if I need documentation for fnm or not. If you have a MacOS computer then I would also like you to try with that - you need Homebrew installed. 
2. I got a bit distracted with the security implications, sorry. Something I was going to have to do with the original Github Action approach was store OAuth creds in a secret to connect to Google. This is not ideal for a few reasons. Since this code will not be run on AppScript, I can use whats known as a Service Account. A Service Account allows auth through a JSON Web Token token which is a lot simpler than oauth,. Security wise, they are not full google accounts - they aren't part of the hackforla.org workspace. Compromise would still be bad but it won't result in someone gaining access to a hackforla.org email account at least. Any resources shared with all hackforla.org account (idk if there are any, but it could happen by accident) are not available to the service account. The main downside of these is the same as the security upside - they are not part of the hackforla.org workspace. The email address that shares the Google Drive folders will look something like `tables@12420032036325.iam.gserviceaccount.com` . If this is acceptable I recommend that, otherwise I'll stick with the current Oauth method.
3. Any update on https://github.com/hackforla/tables/issues/64?

**Comment by ExperimentsInHonesty** [comment]
@hkatzdev In this write-up you don't say why its not ideal.  You just refer to it not being ideal.  Unless I am misunderstanding the write-up.

> 2. I got a bit distracted with the security implications, sorry. Something I was going to have to do with the original Github Action approach was store OAuth creds in a secret to connect to Google. **This is not ideal for a few reasons.** Since this code will not be run on AppScript, I can use whats known as a Service Account. A Service Account allows auth through a JSON Web Token token which is a lot simpler than oauth,. Security wise, they are not full google accounts - they aren't part of the hackforla.org workspace. Compromise would still be bad but it won't result in someone gaining access to a hackforla.org email account at least. Any resources shared with all hackforla.org account (idk if there are any, but it could happen by accident) are not available to the service account. The main downside of these is the same as the security upside - they are not part of the hackforla.org workspace. The email address that shares the Google Drive folders will look something like tables@12420032036325.iam.gserviceaccount.com . If this is acceptable I recommend that, otherwise I'll stick with the current Oauth method.

**Comment by hkatzdev** [comment]
- TL'DR Code is done, documentation and testing is WIP. Blocked by #64, additional reviewer would be greatly beneficial.

Code is written with some comments, need to add typing defs to those comments and finish my Markdown explanations. Did light testing but not enough, also did not have good enough connection to internet for long enough today to set up demo - most of the coding was offline.

Zoom will still be broken if inviting to more than 2 meetings, waiting on fix. I could implement the fix myself if needed but I don't want to cause duplicate work.

Coding it without implementation being fully hidden by App Script has revealed what I believe to be the culprit behind the strange calendar stuff - will implement tests for those, its a race condition like previously discussed. It is fixable, but I believe that we put it in the same category as the Drive race condition and write that the fix is not planned. We are switching away from Calendar soon and it prevents us from using Cloudflare workers - we would have to use Deno Deploy. While I prefer Deno over Node, I'm aware more people have NodeJS experience over Deno. No other API besides drive requires this technical implementation, and Drive has not had this issue in practice due to it only requiring 1 network request for an update compared to 2. Neither Zoom nor Github have this issue, and I don't believe Slack has this issue. Also its pretty easy to fix by unchecking or checking a box in the Tables. Since its not a large issue, I made the decision to stick with Cloudflare and Node.js.

Another thing - we no longer need to have multiple tables with 1 for each CoP. We can still have it, but its no longer needed - thought I would mention that. Async code + Batching FTW here.

Async code also means that 1 service being down won't take the script down with it - its not properly error checked rn but it won't block. Wasn't a huge problem but as we add services this becomes more and more useful. Should also speed up the code, might do a benchmark but honestly the speed of the code is mostly not important. It is only important to try to prevent Drive and Calendar from racing with another worker - if they run quick enough that non run at the same time they can't race ;).

Error Checking isn't implemented here, but we can open an issue to think about that in the future. Its pretty simple the implement - the harder part is figuring out how to display the error. Should there be automatic retries? Should it color a row in the table? Can you even highlight table rows? etc.

- I saw #66 - part of the documentation I'm doing has to do with security. I'll detail key rotation, what the script has access to, how we prevent DoS attacks, how we validate the CoC passphrase, etc. All dependencies are official projects with 1 exception that I have gone through - I'll detail that updating that dep requires a quick diff check for security. The libraries do mean that we mostly don't hand craft fetch requests (zoom is an exception) so things _should_ be handled securely. We have proper environment variables now too.

Most of my testing has been on my personal accounts to prevent messing stuff up, but its ready to migrate to here. Are there any major github changes to this repo happening in the next week? I doubt it but want to verify before changing things.

**Comment by hkatzdev** [comment]
For the additional reviewer - this is the first time I'm doing documentation on this scale. I typically document for myself mostly and not for other developers in general. If anyone is willing to look over my code / docs once they are done for clarity I would greatly appreciate that. Would also like to bounce ideas about the best way to error check and show the error.

**Comment by ExperimentsInHonesty** [comment]
In response to Harrisons message in slack

<details><summary>Details</summary>
<p>

>... unfortunately I made a mistake which is causing it to take longer than expected but its better that this is discovered now rather than later. We are currently doing _no_ input validation beyond that Google Tables allows. This mainly affects Github and Zoom, which have sensitive tokens. Zoom is the most affected due to the way its implemented, but I don't believe the Zoom code is used in production yet. It may be possible for people to add themselves to arbitrary teams on Github though - they would only be a user of the team, but that's still an issue. As long as the forms are only open when needed and someone is watching to make sure no one is doing anything funny, it should be ok. The Website team form is open 24/7, but it only has perms for the website related stuff so only the onboarding form is a concern. To be extra safe I'm commending out the zoom code in appsscript for now.

>I had to rewrite the code with the assumption that user input is not safe, which is what should have been done in the first place - ngl a bit disappointed I'm only catching that now since I do cybersecurity stuff :confused:. Tables form validation is not good enough to prevent an attacker for this. The new codebase is using the official Google and Github libraries, which _should_ escape everything correctly (also adds some cool features like auto retry, which can def be helpful in Google Calendar's case). However, there is no official Zoom library, so I'm trying to be extra cautious there. Additionally, I want to verify the escaping in the library works because zoom's graphql api is a tad strange. To be perfectly honest I'm a bit used to using Typescript to deal with some of this but I am still able to use Javascript as we previously discussed, just causing me to work a bit slower as there are some gotchas between the languages and jsdoc.

>The good news is that the code is able to statically catch mistakes like this now. My code editor gets very angry when I try to use user input without checking it first, which was something not possible with apps script. Its taking longer than expected but there are def improvements to the developer process now. Additionally, I believed I discovered a quick hack thanks to this to hopefully fix the Google Calendar not properly adding users issue (assuming its due to race conditions).

</p>
</details> 

@hkatzdev We cannot rely on the forms being closed manually as a security method.  If we can't secure the code from attacks, we have to find another way of doing it. I will turn off the onboarding form including website for now.  But we need a more robust solution.  Is the problem google tables forms?  Can we use Google Forms (which we were moving to anyway), as a way of doing addition validation?

**Comment by lilykat7** [comment]
@hkatzdev A possible interim solution could be to automatically close the onboarding forms. We could restrict the form opening using a clock/timer to only allow them to be open during onboarding times.

**Comment by hkatzdev** [comment]
> In response to Harrisons message in slack
> Details
> 
> @hkatzdev We cannot rely on the forms being closed manually as a security method. If we can't secure the code from attacks, we have to find another way of doing it. I will turn off the onboarding form including website for now. But we need a more robust solution. Is the problem google tables forms? Can we use Google Forms (which we were moving to anyway), as a way of doing addition validation?

The issue is we have been manually crafting requests for Github and Zoom - this is an oversight due to us being unable to use libraries on Apps Script.

I'm assuming that Google Drive and Calendar library built in autoescapes but I don't know for sure.

This weekend I can do a quick hack to prevent people from having any special characters in their github username. I don't think they are allowed, and if there are no special characters then the code should be fine.
The main concerns is periods because they can craft a custom url (../../) but I recommend all special characters to be safe for now until we switch over to the official Github Library.

Zoom is a bit more complicated, because we send their name as part of the request. I could do a quick hack to escape quotes and backslashes, but we don't use Zoom in production yet so I think I can just comment that out and we can wait until I finish a proper solution before rolling out Zoom invites.

**Comment by ExperimentsInHonesty** [comment]
@hkatzdev yes, we are only allowing them to put in their GitHub handle which can include number, letters and hyphens*.  We only allow a full URL on the LinkedIn field

*Username may only contain alphanumeric characters or single hyphens, and cannot begin or end with a hyphen.

**Comment by hkatzdev** [comment]
> @hkatzdev yes, we are only allowing them to put in their GitHub handle which can include number, letters and hyphens*. We only allow a full URL on the LinkedIn field
> 
> *Username may only contain alphanumeric characters or single hyphens, and cannot begin or end with a hyphen.

Ok, I will enforce this in the code. We currently do not check and tables does not validate it. If they put in something with special characters it will just silently fail, but thats probably what was going to happen before anyways unless they were trying to attack it. 

Hyphens are fine, they can't be used to exploit anything in this case. I'll do `^(\w|-)+$` for appsscript - it allows some invalid Github handles but nothing that poses a security risk.

**Comment by hkatzdev** [comment]
It should be implemented now - I'll do a more thorough test tomorrow.
Zoom code is commented out now too. I looked at the current Drive and Calendar code, but they seem fine since they use the Google library.

There are some small security issues regarding the table with this method but since the table is locked down to trusted contributors I don't find those an issue - someone who has access to change stuff in the table probably has 1password access anyways. Figure I would still mention it though - no one who doesn't need edit access should have it since it still depends on stuff like the Github Teams name to be in proper format (users don't have control over this, only table editors).

**Comment by hkatzdev** [comment]
> @hkatzdev A possible interim solution could be to automatically close the onboarding forms. We could restrict the form opening using a clock/timer to only allow them to be open during onboarding times.

I don't think there is a simple table api to do this, and implementing this would probably take longer than finishing the new code unfortunately.

I do agree that this should still be done manually though. It is _very_ obvious if someone is trying to exploit the system in the tables database. While I hate "security through obscurity", the fact that the code is not publicly available does mean that an attacker would probably be trying many different things with most of them not working, buying time to see spamming in the database and shut the form down.

Unrelated, but something I looked into for both Google Tables and Google Forms is a way to prevent DoS attacks. I was going to bring it up after I finished my current code, but might as well bring it up now. Currently, if someone decides to start spamming the form when its open we have no recourse. The new code's endpoint will be behind Cloudflare (mainly for the free WaF), but Google does not allow you to see the submitter's IP address. This means we have no method to throttle connections based on if an IP is spamming the form. I'm assuming Google Tables and Forms have recaptcha at least but figured I put it out there. If someone can think of a non-ip based solution that doesn't involve clientside javascript (since we can't put js on Google Forms) then I'll be happy to implement it.

**Comment by ExperimentsInHonesty** [comment]
Hi Harrison,

- Please remember to post the message you posted in slack onto this issue.

You can take as much time as you need to recover and catch up on your school work, but we would like for you to turn over all the work you have done so far in whatever state it's in. That means:
- doing a commit to share whatever code changes to the test branch you have been working on.  The code does not have to work, but you do have to commit it.
- creating a document in the Google Drive or adding comments to this issue with whatever notes you have about what you have done, the technology names, etc.  The notes don't have to be perfect, you can keep working on them to improve them (which is why you might want to do a Google Doc), but it's time for you to put your brain dump into our resources, in case something happens to you, and you don't ever come back.

**Comment by ExperimentsInHonesty** [comment]
### Harrison's message from 2024-04-28
Now I have COVID - This has not been the most fun month for me :/.
Going to have to isolate for a week + dealing with symptoms (luckily its only like a moderate flu and not more severe).
This means that I haven't been able to work on stuff this week which I planned to. Also means I'm going to end up a week behind with school and need to catch up yet again
I think we are going to have to place the test environment project on ice for now - I can't put the final remaining time needed for it, either I've been too sick to focus on it or I've been forced to catch up with school.
I can probably only spend 1 hour a week for the next 2 weeks at least on tables - that is enough for many quick changes in apps script though. Will probably not attend meeting this week unless throat improves, idk about next week depends on when teacher reschedules my tests.

Sorry for another update like this - on the good news the new progress on the code has been really good (which is why I'm a bit sad to put it on hold). It is safe from future architecture changes - if tables shuts down, it can easily be fit to sheets, airtable, custom, etc. It can handle basically as many people as we would ever need without timing limits. It is compatible with at least 3 free cloud providers, so if we ever have to switch in the future its no hassle. Some of the invite code could theoretically be reused in VRMS if needed - its all Node compatible code. Libraries include the auto-retry for less failures & good validation. It also uses Cloudflare's free WAF on the backend to prevent some spamming attacks, though without IP info (which isn't given by tables) its not too effective that it - it does prevent obvious injections style attacks in case something is overlooked though.

I think from reading the above its easy to see where I went wrong - I kinda exploded the scope of my issue. I got a bit too carried away with documenting stuff, and whenever I had a question and thought there was a better way I immediately explored what the better way was. The main issue I got was that it was hard to make working test environment automation that was both easy to understand and free. I could hack together a free testing automation tool, but you would need some experience with tools people don't typically use. Making tables cost money is out of scope of this project. I am happy with what I've achieved so far, but it still has solid hours of consecutive work before its finally complete.

### Harrison's message from 2024-05-08
I will upload my code by next week. Covid is gone now thankfully, but I got a really bad cough afterwards - thankfully it seems to be getting better but I don't think my voice can handle a meeting.
Some good news and less good news - I figured out most of the problems when I had COVID except Google Calendar, so the code I upload will be mostly complete. Google Calender has the same problem as the existing code, but due to how the new code is written it may be worse - no way to know without testing. I highly dislike this but there is no free simple solution to this problem - I will detail the reasons in a doc. The less good news is that my laptop broke earlier this week - it's a battery / charging issue, so nothing is lost. I have several backups of the code so that's all good, but my git apps, accounts like onepassword, and some documentation notes are on the laptop so I need to order a pcie to USB or smth to get my data off (edited)

**Comment by hkatzdev** [comment]
Cough is still not great unfortunately - seems to come and go in cycles.

I wasn't able to work on this last week due to that - additionally I only got the part to get my data off a day or two ago.
However, I did upload a VSCode Workspace for now at least to Google Drive at https://drive.google.com/file/d/1xNNByAfG09dAaPDBZo5f3CxgEXI4xGCX/view - this is all the code including git ignore files. It includes some config, style, run settings, comments, etc. I will try to upload to git w/ documentation by next week - however at least all the code is there.

This file is openable in VSCode or VSCodium.

Hoping to have git upload and documentation by next week, however I should at minimum have it done by the 31st since that's after my finals. 

Was hoping to attend meeting today but still not able to unfortunately.

> Hi Harrison,
> 
>     * Please remember to post the message you posted in slack onto this issue.
> 
> 
> You can take as much time as you need to recover and catch up on your school work, but we would like for you to turn over all the work you have done so far in whatever state it's in. That means:
> 
>     * doing a commit to share whatever code changes to the test branch you have been working on.  The code does not have to work, but you do have to commit it.
> 
>     * creating a document in the Google Drive or adding comments to this issue with whatever notes you have about what you have done, the technology names, etc.  The notes don't have to be perfect, you can keep working on them to improve them (which is why you might want to do a Google Doc), but it's time for you to put your brain dump into our resources, in case something happens to you, and you don't ever come back.

**Comment by ExperimentsInHonesty** [comment]
@hkatzdev I moved the file you uploaded to the Tables drive (instead of the mydrive of the tables email).  I tried to open it in VS code, but all it did was show me my own files, not any you created. 

[VS workspace file](https://drive.google.com/file/d/1xNNByAfG09dAaPDBZo5f3CxgEXI4xGCX/view )

**Comment by hkatzdev** [comment]
I misunderstood the workspace file, my bad. I uploaded a zip instead @ https://drive.google.com/file/d/1yL4mEqE7nRbC2B2D-B6SjgtnUdV8MmLE/view.
Warning that its a bit of a mess because it has some failed test / old code in the archive that isn't cleaned up - planning to clean that up and make a proper gitignore ASAP. The important code bits are the .js files in src - any other files anywhere else (root, other subfolders, .ts, .json, etc) are unimportant and either test stuff, old remnants, or style config files.

> @hkatzdev I moved the file you uploaded to the Tables drive (instead of the mydrive of the tables email). I tried to open it in VS code, but all it did was show me my own files, not any you created.
> 
> [VS workspace file](https://drive.google.com/file/d/1xNNByAfG09dAaPDBZo5f3CxgEXI4xGCX/view)

**Comment by hkatzdev** [comment]
Will finish git upload and notes by next meeting but I do want to post the main takeaways.

Ultimately, automated test environment was unable to fulfill the main three requirements originally set: free, simple, and easily maintainable.

I set up trying to solve this using serverless technology since thats how AppScript works, but standard tech has the potential to cause more Google Calendar bugs. It could also interfere with other future services we may want to add. We need a built in Queue service in order to ensure we can control the execution speed and concurrency of the API requests.

There are three main services that I evaluated: Cloudflare, Deno Deploy, and AWS. All three of these services failed a requirement. Cloudflare is simply and easy to maintain, but costs $5 a month. Deno Deploy is free, but requires too many workarounds to leave the code understandable*. AWS is free and leaves the code simple, but was already turned down at a prior meeting due to not being easily maintainable - I could tell from writing just a bit of AWS code that a lot of documentation would be needed, and this project doesn't have a devops role.

There are still some advantages from switching, but I am hesitant to recommend changing services if it is not nessassary. We could remove timeout restrictions, use official libraries, have better failsafes, etc. However, the scope of the tables project is still small - we don't add new services too often, we can manually validate input, we don't mind having to recheck a box if it doesn't work the first time, etc. I've definitely got some ideas on how to improve the AppScript code all around from this. 

Ultimately, I'm going to just write some guides on how to set up a testing environment. It is kinda convoluted to do manually - maybe I can investigate some helper scripts, but I will do that after the guides are written.

Some circumstances that could change this:
1) If the forms ever get spammed, switching away from AppScript allows us to do IP based rate limiting. This has not been a problem, but I mention it since I think I heard Hack For LA had a zoom bombing problem a while back.
2) Deno releases 2.0. The 2.0 release should be feature complete enough to be usable for our case. I don't expect that for months though. Once it is released, this issue can be reopened.
3) We have an influx of users. If we ever start onboarding 100s of users (!!!) the above options can easily handle the load - they can probably handle thousands of requests easily and for free. This is more relevant if the code is reused for something and we want it all on the same infrastructure.
4) We decide to remove a requirement. Unless Hack for LA is already paying for Cloudflare or has extra dev ops people I don't see this as being feasible.
5) Hack for LA has a free server (that isn't AWS). Highly doubt, but if so it would be pretty trivial to turn this into serverful code. Won't be as fast as the above mainstream options but would still be a big improvement over AppScript and would allow easy testing environment automation.

I'm sure there are more which I'll type up into a full design doc. Finished my last final yesterday, was hoping I could finish the documentation yesterday too but I underestimated how tired I would be once I finished it.

Cloudflare Github Action: https://github.com/cloudflare/wrangler-action

**Comment by hkatzdev** [comment]
Sorry for the delay - final code will be pushed by Thursday night. Last week and this Monday I has to go into school to sort smth out, so only could work on this yesterday and today.

When I was experimenting with the three different services, I wrote the code in a very object oriented class based way. This made it extremely simple to test out different internal implementations of stuff. However, now that we decided on Cloudflare, it was way over complicated and inefficient. I forgot I had to undo those changes last meeting - thats mostly done, 3 out of the 5 files are complete and halfway through the 4th. The 5th file did not involve classes and should be much shorter to fix - just need to change some variable names. That part should be done by tonight.

With Cloudflare, we can also somewhat isolate the individual environment variables. This means that the Github related code doesn't have access to our Google Keys, Google Calendar code doesn't need Google Drive access, and the initial webhook that receives user input doesn't need access to _any_ environment variables. This way, if one part of the code is compromised, its possible to limit the impact. This also has a nice side affect of allowing us to deploy features and bug fixes for some services without affecting others - currently, all the code runs as one application. With this change, if a breaking change is accidentally pushed to the Drive for example, Calendar and Github will be unaffected.

This does mean I need to finish setting up the files before pushing though - part of the setting up is wrangler config. Wrangler is a cloudflare app that allows us to run the code locally - right now, even if I had the completed files without the config you can't do much besides just looking at them. I've done a small test on part of it and it seems to be easy enough, but I want to make sure it works for the full project and add a tiny bit of documentation on running it - that should be done by tmr night.

Overall it is looking good, it looks much more maintainable to me. I uploaded a copy of what I currently have to git, but again want to mention it is currently incomplete. It does have the general project structure, apps needed, and google drive + google calendar files mostly done though.

P.S. The unknown commit is me, I haven't set up git properly on this computer yet.

**Comment by ExperimentsInHonesty** [comment]
@hkatzdev Please provide update
1. Progress: "What is the current status of your project? What have you completed and what is left to do?"
2. Blockers: "Difficulties or errors encountered."
3. Availability: "How much time will you have this week to work on this issue?"
4. ETA: "When do you expect this issue to be completed?"
5. Pictures or links* (if necessary): "Add any pictures or links that will help illustrate what you are working on."

* remember to add links to the top of the issue if they are going to be needed again.

**Comment by ExperimentsInHonesty** [comment]
Also, there are a bunch of script files in tables@hackforla.org but it's unclear which ones are in use, etc. Please get back to us with a compressive list.

**Comment by ExperimentsInHonesty** [comment]
@hkatzdev ???

**Comment by hkatzdev** [comment]
> @hkatzdev Please provide update
> 
>     1. Progress: "What is the current status of your project? What have you completed and what is left to do?"
> 
>     2. Blockers: "Difficulties or errors encountered."
> 
>     3. Availability: "How much time will you have this week to work on this issue?"
> 
>     4. ETA: "When do you expect this issue to be completed?"
> 
>     5. Pictures or links* (if necessary): "Add any pictures or links that will help illustrate what you are working on."
> 
> 
>     * remember to add links to the top of the issue if they are going to be needed again.

1 and 3 are related - was unable to work on much of anything for the past 4 weeks due to outside circumstances unfortunately. However I am once again free to work on this and can spend 5 hours a week for the next couple of weeks, then 5 hours a week consistently.

2. Not a blocker but something I'm trying to figure out is database testing in practice. Will give larger update next week, may make another issue for that. No hard blockers though will need cloudflare account soon to do full testing / deployment,

4. Hoping by next week since I have time again, was working on it a tad the past couple of days and it seems nearly ready to deploy, though documentation is still lacking and I need to upload it to git,

**Comment by hkatzdev** [comment]
> Also, there are a bunch of script files in [tables@hackforla.org](mailto:tables@hackforla.org) but it's unclear which ones are in use, etc. Please get back to us with a compressive list.

https://script.google.com/home/projects/1pRLXW3uUjSvePgxldfDzvix43AuiYaZ7aC3T1gVh9q7PXoFCdaB1sKau is used for most, and https://script.google.com/home/projects/19Rxe3GIAGmesJDLhjTfGn88bhCARyrlkExpQVb7HnIT85nNvdeLH4SnC is used just for Github teams. Zoom is unused due to security rn, everything else is old or test code.

**Comment by hkatzdev** [comment]
Can't attend meeting today but will be able to attend regularly starting next week - I cleared my Wednesdays from 2:30-3:30 at least until December.

**Comment by hkatzdev** [comment]
To expand on https://github.com/hackforla/tables/issues/64#issuecomment-2302924942, Zoom was removed from the MVP of this due to #72. Zoom does _not_ provide an official node library, which means the code written for zoom must be triple checked to ensure no exploits. This comes up with escaping - I'm not super familiar with escaping graphql stuff and those possible injections, and that should be investigated more before deploying zoom related stuff. This came up before when it turned out we do not properly sanitize user input which could allow for some injection attacks.

Google and Github provide official libraries that properly sanitize input or at least prevent harmful effects and is not an issue in this regard.

**Comment by hkatzdev** [comment]
Drive and Github code is ready for testing - ideally we can set up the cloudflare account so I can test some wrangler configs and get it ready to push

Calendar code needs a little more time - mainly its refactoring it to match the style of the Drive and Github. I went through a few designs alternating between very extendable but too complex to short but hacky before finding a way to write it in a readable way.

The project is set up so they can all run on separate instances - that means no sharing Google api keys with the zoom code and vice verse. This introduces some complexity, but I'm mostly managed to hide that away. A new dev can just cd to the part they want to work on and its like a normal node project.

Additionally, I'm using an old deno convention of a deps.js file. Typescript was previously eliminated, but types have helped me catch many small errors. The compromise is jsdoc. JSDoc is great for functions - it forces functions to be properly annotated. However, despite recent improvements, importing types is very verbose - a third of the file would be imports. Moving this to a separate file allows people who aren't messing with types to ignore it - as an additional bonus, it makes it look more familiar to appscript devs. Also allows separating some data structures for the code.

Automatic retry on failure is implemented - currently its on any failure with default 3 times settings. So a request that times out could still succeed later. While the new calendar code should theoretically not have the concurrent update bug, Google Calendar is so unreliable that its good to have this as a backup.

Out of scope for this issue but 2 other interesting things are (a) it is trivial to detect when smth fails now. A better way for queues is to say not proven to succeed, but its the same for our purposes sense running the same operation twice is not an issue. I don't know how we can display this on the table yet. The other is (b) it reworks github to rely on the user's id now. It gets the id from the username so no change to the form necessary, but it would be interesting to see if this can be saved to the table in case someone changes their username - you can't change your github id.

**Comment by hkatzdev** [comment]
https://dash.cloudflare.com/dd2087430414abfb7bdff7b6e3c46a7e/workers/plans/purchase

**Comment by hkatzdev** [comment]
I made a tables@hackforla.org user that does not have billing perms - should be safe to move the cloudflare@hackforla.org account to an admin vault before adding billing info. tables prob has too many perms but I wanted to be lenient for now at least since its the only cloudflare user - later on we can look at reducing its scope further.

**Comment by hkatzdev** [comment]
![Screenshot 2024-09-25 132937](https://github.com/user-attachments/assets/37a2c4fb-c967-447f-8182-43aeeb929c72)
I've done as much local testing as possible, unfortunately I'm blocked now until we have a paid workers plan

**Comment by ExperimentsInHonesty** [comment]
Upgraded account

**Comment by ExperimentsInHonesty** [comment]
Please commit to github!!!!!!!!!!!!

**Comment by hkatzdev** [comment]
Testing revealed 2 errors, 1 is fixed the other is relatively minor.
Haven't uploaded to git yet because cloudflare wants access over the entire repo, not just a branch - was wondering if its ok if I make a separate git repo (tables-staging) to be safe - dont want to accidentally erase anything here.
Need to finish docs and config, but everything seems to be working besides the identified issues

**Comment by ExperimentsInHonesty** [comment]
- created a new repo https://github.com/hackforla/tables-test-platform
- added it to the tables team

Harrison will use that to use that to connect to cloudflare

**Comment by ExperimentsInHonesty** [comment]
@hkatzdev Please provide update
1. Progress: "What is the current status of your project? What have you completed and what is left to do?"
2. Blockers: "Difficulties or errors encountered."
3. Availability: "How much time will you have this week to work on this issue?"
4. ETA: "When do you expect this issue to be completed?"
5. Pictures or links* (if necessary): "Add any pictures or links that will help illustrate what you are working on."

* remember to add links to the top of the issue if they are going to be needed again.

**Comment by ExperimentsInHonesty** [comment]
@hkatzdev Please give an update on this issue, and let us know how we can move on from where you are now, so that we don't continue to be stuck waiting for a reply from you.

**Comment by hkatzdev** [comment]
I apologize, been packed with transfer applications and school work. Will give a proper update on the 17th. This issue shouldn't be blocking any existing issues though since additional code is easy enough to convert. If there are specific blockers by the issue please let me know and I will prioritize resolving those.

**Comment by Reltre** [comment]
@hkatzdev We're putting a move to Cloudflare on hold for now. Many of the security concerns mentioned in your write-up have been sorted out for the purposes of Tables team. It isn't entirely off the table forever, but it isn't needed right now.

**Comment by Reltre** [comment]
Closing this issue for now.
